src/librbd/mirror/snapshot/PromoteRequest.cc

```diff
@@ -0,0 +1,392 @@
+  ldout(cct, 20) << dendl;
+
+  auto ctx = create_context_callback<
+    PromoteRequest<I>, &PromoteRequest<I>::handle_refresh_image>(this);
+  m_image_ctx->state->refresh(ctx);
+}
+
+template <typename I>
+void PromoteRequest<I>::handle_refresh_image(int r) {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(cct) << "failed to refresh image: " << cpp_strerror(r) << dendl;
+    finish(r);
+    return;
+  }
+
+  if (!util::can_create_primary_snapshot(m_image_ctx, false, m_force,
+                                         &m_rollback_snap_id)) {
+    if (!m_force) {
+      lderr(cct) << "cannot promote" << dendl;
+      finish(-EINVAL);
+      return;
+    }
+    ldout(cct, 0) << "force promoting without rollback to known state" << dendl;
+  } else if (m_rollback_snap_id == CEPH_NOSNAP) {
+    create_promote_snapshot();
+    return;
+  }
+
+  create_orphan_snapshot();
+}
+
+template <typename I>
+void PromoteRequest<I>::create_orphan_snapshot() {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  auto ctx = create_context_callback<
+    PromoteRequest<I>,
+    &PromoteRequest<I>::handle_create_orphan_snapshot>(this);
+
+  auto req = CreateNonPrimaryRequest<I>::create(m_image_ctx, "", CEPH_NOSNAP,
+                                                {}, nullptr, ctx);
+  req->send();
+}
+
+template <typename I>
+void PromoteRequest<I>::handle_create_orphan_snapshot(int r) {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(cct) << "failed to create orphan snapshot: " << cpp_strerror(r)
+               << dendl;
+    finish(r);
+    return;
+  }
+
+  list_watchers();
+}
+
+template <typename I>
+void PromoteRequest<I>::list_watchers() {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  auto ctx = create_context_callback<
+    PromoteRequest<I>,
+    &PromoteRequest<I>::handle_list_watchers>(this);
+
+  m_watchers.clear();
+  auto flags = librbd::image::LIST_WATCHERS_FILTER_OUT_MY_INSTANCE |
+               librbd::image::LIST_WATCHERS_MIRROR_INSTANCES_ONLY;
+  auto req = librbd::image::ListWatchersRequest<I>::create(
+    *m_image_ctx, flags, &m_watchers, ctx);
+  req->send();
+}
+
+template <typename I>
+void PromoteRequest<I>::handle_list_watchers(int r) {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(cct) << "failed to list watchers: " << cpp_strerror(r)
+               << dendl;
+    finish(r);
+    return;
+  }
+
+  if (m_watchers.empty()) {
+    acquire_exclusive_lock();
+    return;
+  }
+
+  wait_update_notify();
+}
+
+template <typename I>
+void PromoteRequest<I>::wait_update_notify() {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  ImageCtx::get_timer_instance(cct, &m_timer, &m_timer_lock);
+
+  std::lock_guard timer_lock{*m_timer_lock};
+
+  m_scheduler_ticks = 5;
```

Comment:
What's the purpose of this register/unregister loop? Was it just easier to implement for the state machine?

Reply:
Yes, it was just easier to implement this way.